File: src/kotlin/app/src/main/kotlin/ru/renev/adventofcode2021/Day6.kt
package ru.renev.adventofcode2021

class Day6 {
    data class LanternsGroup(val timer: Int, val count: Long) {
        fun nextGroup() : List<LanternsGroup> {
            if (timer != 0) {
                return listOf(LanternsGroup(timer - 1, count))
            }

            return listOf(
                LanternsGroup(6, count),
                LanternsGroup(8, count)
            )
        }
    }

    fun part1() : Long {
        println("Day 6 part 1")

        return calculateCount(80)
    }

    fun part2() : Long {
        println("Day 6 part 2")

        return calculateCount(256)
    }

    private fun calculateCount(days: Int) : Long {
        var lanterns = parseInput().groupBy { it }.
            map { LanternsGroup(it.key, it.value.count().toLong()) }.
            toList()

        repeat(days) { _ ->
            lanterns = lanterns.flatMap { it.nextGroup() }.
            groupBy { it.timer }.
            map { grp -> LanternsGroup(grp.key, grp.value.sumOf { it.count }) }.
            toList()
        }

        return lanterns.fold(0L) { a, b -> a + b.count }
    }

    private fun parseInput(): List<Int> {
        return Day6Input.input.split(",").map { it.toInt() }
    }
}

File: src/dotnet/AdventOfCode2021/AdventOfCode2021.Solutions/Day6.cs
namespace AdventOfCode2021.Solutions;

public partial class Day6 : IDay
{
    private record class LanternsGroup(int Timer, long Count)
    {
        public IEnumerable<LanternsGroup> GetNextGroup()
        {
            if (Timer != 0)
            {
                yield return new LanternsGroup(Timer - 1, Count);
                yield break;
            }

            yield return new LanternsGroup(6, Count);
            yield return new LanternsGroup(8, Count);
        }
    }

    public long CalculatePartOne()
    {
        return CalculateCount(80);
    }

    public long CalculatePartTwo()
    {
        return CalculateCount(256);
    }

    public int DayNumber => 6;

    private static long CalculateCount(int days)
    {
        var lanterns = ParseInput().GroupBy(val => val)
            .Select(grp => new LanternsGroup(grp.Key, grp.Count()))
            .ToArray();

        foreach (var i in Enumerable.Range(0, days))
        {
            lanterns = lanterns.SelectMany(l => l.GetNextGroup())
                .GroupBy(g => g.Timer)
                .Select(grp => new LanternsGroup(
                    grp.Key,
                    grp.Select(g => g.Count).Aggregate((v1, v2) => v1 + v2))
                    )
                .ToArray();
        }

        return lanterns.Select(l => l.Count).Aggregate((v1, v2) => v1 + v2);
    }

    private static int[] ParseInput()
    {
        return Input.Split(",").Select(int.Parse).ToArray();
    }
}
